Use pgBackRest TLS server

cbandy opened the pull request:
**Checklist:**

 <!--- Make sure your PR is documented and tested before submission. Put an `x` in all the boxes that apply: -->
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [x] Have you updated or added documentation for the change, as applicable?
 - [ ] Have you tested your changes on all related environments with successful results, as applicable?
   - [x] Have you added automated tests?


**Type of Changes:**

 <!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
 - [x] New feature


**What is the current behavior (link to any open issues here)?**

When a PostgresCluster is configured to store backups on a PVC, we start a dedicated repository host to make that PVC available to all PostgreSQL instances in the cluster. The repository host and each PostgreSQL instance run an SSH server to encrypt, authenticate, and authorize connections between pods.


**What is the new behavior (if this is a feature change)?**
- [x] Breaking change (fix or feature that would cause existing functionality to change)

The repository host and each PostgreSQL instance run a pgBackRest server to encrypt, authenticate, and authorize connections over mTLS. A configuration sidecar runs alongside the server to detect and load changes to configuration and certificates. The directory structure:

```console
$ bash -c 'ls -ldH /etc/pgbackrest /etc/pgbackrest/* /etc/pgbackrest/*/* /etc/pgbackrest/*/*/*'
drwxrwxr-x. 1 postgres postgres  34 Dec  4 02:51 /etc/pgbackrest
drwxrwsrwt. 3 root     postgres 160 Dec  4 02:51 /etc/pgbackrest/conf.d
-rw-r--r--. 1 root     postgres  10 Dec  4 02:51 /etc/pgbackrest/conf.d/config-hash
-rw-r--r--. 1 root     postgres 928 Dec  4 02:51 /etc/pgbackrest/conf.d/pgbackrest_repo.conf
drwxr-sr-x. 2 root     postgres 100 Dec  4 02:51 /etc/pgbackrest/conf.d/~postgres-operator
-rw-r--r--. 1 root     postgres 696 Dec  4 02:51 /etc/pgbackrest/conf.d/~postgres-operator/client-tls.crt
-rw-r-----. 1 root     postgres 227 Dec  4 02:51 /etc/pgbackrest/conf.d/~postgres-operator/client-tls.key
-rw-r--r--. 1 root     postgres 492 Dec  4 02:51 /etc/pgbackrest/conf.d/~postgres-operator_server.conf
-rw-r--r--. 1 root     postgres 583 Dec  4 02:51 /etc/pgbackrest/conf.d/~postgres-operator/tls-ca.crt
drwxrwsrwt. 3 root     postgres 120 Dec  4 02:51 /etc/pgbackrest/server
-rw-r--r--. 1 root     postgres 928 Dec  4 02:51 /etc/pgbackrest/server/server-tls.crt
-rw-r-----. 1 root     postgres 227 Dec  4 02:51 /etc/pgbackrest/server/server-tls.key
```

**Other Information**:

Issue: [sc-12859]

tjmoore4 commented on internal/controller/postgrescluster/pgbackrest.go:
```diff
@@ -561,32 +560,35 @@ func (r *Reconciler) generateRepoHostIntent(postgresCluster *v1beta1.PostgresClu
+	// - https://docs.k8s.io/concepts/workloads/pods/pod-lifecycle/#restart-policy
+	repo.Spec.Template.Spec.RestartPolicy = corev1.RestartPolicyAlways
+
+	// The pgBackRest TLS server handles client connections in detached/orphaned
```
:pick:  I understand that there are cases where we'll have an orphaned child process, so we want to use the Kubernetes pause process to reap them, but I'm missing how the pause process is actually being used. More clarification here might be helpful.

cbandy replied:
Setting `spec.shareProcessNamespace: true` is what sets the pause process to PID 1.

My only idea is to say that in the comment, but that's also what the line of code says. Any ideas how to phrase it?

tjmoore4 replied:
Looking at this again, I think my confusion was that the `shareProcessNamespace` setting BOTH makes processes visible across containers in a Pod AND results in the PID 1 `/pause` process.

I think adding another link would make this clearer, but then it would be repetitive, i.e.
```
// The pgBackRest TLS server handles client connections in detached/orphaned
// child processes. Use Kubernetes' pause process as PID 1 to reap them.
// - https://github.com/pgbackrest/pgbackrest/blob/release/2.36/src/command/server/server.c#L51
// - https://kubernetes.io/docs/tasks/configure-pod-container/share-process-namespace/#understanding-process-namespace-sharing
//
// The pgBackRest TLS server must be signaled when its configuration or
// certificates change. Let containers see each other's processes.
// - https://docs.k8s.io/tasks/configure-pod-container/share-process-namespace/
```
So I'm not sure that's the way to go. It may not be confusing to others, so perhaps it's fine as is.

cbandy replied:
Now that I read those links again, they don't mention child processes or reaping. I see if I can add some language around that (undocumented?) aspect of the pause process.

tjmoore4 replied:
The updated comment and the additional link really help :100:

tjmoore4 commented on internal/pgbackrest/reconcile_test.go:
```diff
@@ -353,6 +536,267 @@ func TestAddSSHToPod(t *testing.T) {
+		// Only Containers and Volumes fields have changed.
+		assert.DeepEqual(t, pod, *out, cmpopts.IgnoreFields(pod, "Containers", "Volumes"))
+
+		// The TLS server is added while other containers are untoched.
```
:pick: 
```suggestion
		// The TLS server is added while other containers are untouched.
```

tjmoore4 reviewed (approved):
:+1: Looks good, just a few questions/nitpicks, but no show-stoppers.

cbandy commented on internal/pgbackrest/config.go:
```diff
@@ -200,11 +221,17 @@ mv "${pgdata}" "${pgdata}_bootstrap"`
 	globalConfig map[string]string,
 ) iniSectionSet {
 
+	// TODO(cbandy): pass a FQDN in already.
```
I noticed that `serviceName` and `serviceNamespace` are used only to build the FQDN. The caller is already passing in `repoHostName`, so it seemed reasonable that they pass the exact name they want in the configuration.

That would also push this `context.Background()` out to somewhere that probably does have a Context (controller, maybe?).

tjmoore4 replied:
:+1: gotcha, tracking now, thanks.

cbandy replied:
These functions are called exactly once by `Reconciler.reconcilePGBackRestConfig` which has some SSH in its signature. Knocking out these TODOs should be quick, but I'll wait until after we completely remove SSH.

cbandy commented on internal/pgbackrest/reconcile.go:
```diff
@@ -81,58 +85,149 @@ func AddRepoVolumesToPod(postgresCluster *v1beta1.PostgresCluster, template *cor
+		{Key: CMRepoKey, Path: CMRepoKey},
+		{Key: ConfigHashKey, Path: ConfigHashKey},
+		{
+			Key:  serverConfigMapKey,
```
The shape of the literal? Like this?

```go
{
  Key:  ConfigHashKey,
  Path: ConfigHashKey,
},
```

tjmoore4 replied:
Yes, either that, or make them all single line, as in
```
	configmap.ConfigMap.Items = []corev1.KeyToPath{
		{Key: CMRepoKey, Path: CMRepoKey},
		{Key: ConfigHashKey, Path: ConfigHashKey},
		{Key: serverConfigMapKey, Path: serverConfigProjectionPath},
	}
```
Either option is fine with me.

cbandy commented:
Rebased and addressed comments. Had to push twice, so here's the diff:

https://github.com/CrunchyData/postgres-operator/compare/eee2777a95b131471f852ac83e1a99750d3d0b46..121808dc1110ac734615e2ba9aee29648324d770

tjmoore4 reviewed (approved):
Updates look good :+1:
